text/0000-manifest-list-buildpacks.md

```diff
@@ -0,0 +1,159 @@
+
+The spec for [`buildpack.toml`](https://github.com/buildpacks/spec/blob/main/buildpack.md) makes no mention of `os` or `architecture`, which means the adding those fields to entries in the `metadata.dependencies` list seems like a perfectly reasonable option. [This](https://github.com/paketo-buildpacks/rfcs/blob/decouple-dependencies/text/0000-decouple-dependencies.md) rfc proposes adding `os` and `arch` fields to the versions in the dependency table. I suggest sticking with the same field names that are used in the manifest image (`os` and `architecture`) for consistency, but the general idea is the same.
+
+Here is an example of what the dependencies would look like.
+
+```toml
+  [[metadata.dependencies]]
+    cpe = "cpe:2.3:a:golang:go:1.18.4:*:*:*:*:*:*:*"
+    id = "go"
+    licenses = ["BSD-3-Clause"]
+    name = "Go"
+    stacks = ["*"]
+    strip-components = 1
+    version = "1.18.4"
+    os = "linux"
+    architecture = "amd64"
+    purl = "pkg:generic/go@go1.19?checksum=c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5&download_url=https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+    sha256 = "c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5"
+    source = "https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+    source_sha256 = "c9b099b68d93f5c5c8a8844a89f8db07eaa58270e3a1e01804f17f4cf8df02f5"
+    uri = "https://go.dev/dl/go1.18.4.linux-amd64.tar.gz"
+
+
+  [[metadata.dependencies]]
+    cpe = "cpe:2.3:a:golang:go:1.18.4:*:*:*:*:*:*:*"
+    id = "go"
+    licenses = ["BSD-3-Clause"]
+    name = "Go"
+    stacks = ["*"]
+    strip-components = 1
+    version = "1.18.4"
+    os = "linux"
+    architecture = "arm64"
+    purl = "pkg:generic/go@go1.19?checksum=35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f&download_url=https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+    sha256 = "35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f"
+    source = "https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+    source_sha256 = "35014d92b50d97da41dade965df7ebeb9a715da600206aa59ce1b2d05527421f"
+    uri = "https://go.dev/dl/go1.18.4.linux-arm64.tar.gz"
+```
+
+### Github actions and workflows
+
+I haven't looked at github workflows for all buildpacks, but I believe [jam](https://github.com/paketo-buildpacks/jam) is being used to package buildpacks. I did see skopeo being used to push images ([cpython](https://github.com/paketo-buildpacks/cpython/blob/main/.github/workflows/push-buildpackage.yml#L62)).
+
+My proposal is to update jam to create buildpacks for amd64 and arm64. I haven't used jam yet, so I don't have specific code changes for it and would appreciate thoughts on how this could be added.
```

Comment:
As I'm thinking more about this, it's not obvious to me what, if any, changes are required to `jam pack` to support building multi-arch buildpacks. `jam pack` doesn't compile the buildpack; it just packages it. It also doesn't create OCI images - `pack buildpack package` does this.

Expanding on this in more detail:

The current workflow to create a buildpack is:

1. Compile the `build` and `run` binaries (e.g. `go build`).
1. Run either `jam pack` (for `packit` buildpacks) or `packager` (for `libbuildpack` buildpacks). This primarily creates a tgz of the buildpack files. For `packit` buildpacks, this also does some extra steps like setting the buildpack version and optionally downloading dependencies. I'm not sure what other steps `packager` takes.
1. `pack buildpack package` to turn that `tgz` into an OCI image.

Currently, `jam pack` creates a single tarball of a single buildpack with no concept of multi-architecture. You can optionally filter dependencies by `--stack` to only include the stack-specific dependencies in the resultant metadata (and downloaded dependencies), but given that stacks can be multi-architecture, this isn't really sufficient to filter dependencies by architecture.

It's not really obvious to me what it would mean for `jam pack` to support packaging up a buildpack in a multi-architecture way. It doesn't compile the binaries, and it doesn't produce OCI images. Filtering the dependencies by architecture would be an optimization to reduce metadata (and resultant buildpack size if downloading dependencies), but it's not a necessary step for creating the buildpack.

I suppose you could make modifications to `jam pack` to support packaging multiple versions of `build` and `run` for each target architecture and create a wrapper around them to invoke the correct one at run-time. But it seems just as easy to invoke `jam pack` multiple times - once per architecture - to create architecture-specific tarballs.

I don't know if `pack buildpack package` already supports multi-arch images. I took a quick look at the code, and it [only seems to be aware of `platform`](https://github.com/buildpacks/pack/blob/12bc0e241895c9fec3940164ea9771f7ea4f5e53/buildpackage/config_reader.go#L22), which [only contains `OS`](https://github.com/buildpacks/pack/blob/12bc0e241895c9fec3940164ea9771f7ea4f5e53/pkg/dist/dist.go#L38) (i.e. `linux` or `windows`). This leads me to think either we'd have to make changes to `pack`, or we'd have to create separate docker images for each architecture, and bundle them together. Perhaps that's what `docker buildx imagetools create` does - I'm not sure. Perhaps @dmikusa or @dashaun have more info on how to build multi-arch buildpacks in a single manifest?

Reply:
You are correct that `pack` does not support multi-arch yet, and the new changes being introduced will allow users to create manifest images after they have already been built.

[This](https://github.com/jericop/cnb-builder/blob/01a194b802311e09c9acffb53470a374f8d618b3/.github/actions/multi-arch-builder-test/action.yml) is an example of creating multi-arch builders. It creates and publishes arch-specific tags, but it is meant for demonstration purposes and testing so it gives an idea.

Creating multi-arch buildpacks should be simpler and may not even require a local registry because no other images are pulled in. That should allow you to create the arch-specific buildpack images locally and then create a manifest from there and push the manifest.

Let's say you compile a buildpack for amd64 in one folder `amd64` and arm64 in another folder `arm64`. You then create local buildpack images for each arch such as `mybuildpack:0.0.1-amd64` and `mybuildpack:0.0.1-arm64`. Then you can just create and push a manifest image as shown below and it would work on both architectures.

`docker buildx imagetools create -t mybuildpack:0.0.1 mybuildpack:0.0.1-amd64 mybuildpack:0.0.1-amd64`

There is one issue that needs to be considered though. At this time pack seems to set the architecture to `amd64`. I had to work around this issue when creating the multi-arch builder [here](https://github.com/jericop/cnb-builder/blob/01a194b802311e09c9acffb53470a374f8d618b3/.github/workflows/build-and-publish.yml#L27), but I just wanted to call this out.

Reply:
Yes, that is exactly what I was thinking. I don't think we need to make any changes to `jam pack` to achieve this.

It would look something like the following:

1. for arch in [amd64, arm64]
    1. compile buildpack executables for target architecture via `go build`; put them in arch-specific temp directory
    1. package buildpack with `jam pack` into arch-specific temp directory
    1. create arch-specific OCI image via `pack buildpack package`, passing in contents of arch-specific temp directory
    1. upload image to local docker daemon/container-runtime/engine (whatever the correct terminology is here) with arch-specific tag

1. create single multi-arch image via `docker builds imagetools create` and the arch-specific tagged images previously uploaded to local daemon
1. publish single image to relevant registries

Reply:
More of an implementation detail, not sure it needs to go into the RFC, but `skopeo` won't work with multi-arch images so `jam` may need some tooling for that. Not sure if the commands I added for publishing the stacks will work with buildpacks as well? Maybe 🤔 

On the pipeline-builder side, we use `crane` and it does not work with multi-arch images either. We'll need a similar tool. It could be a good chance for both pipelines to remove some 3rd party tools and rally around `jam` as our solution.

Reply:
More thoughts. Pack is adding [a `manifest` command for generating manifest images](https://github.com/buildpacks/pack/pull/1705). This is part of their first step towards multi-arch support. If it is ready in time, it could be something we utilize instead of `docker`. Not sure what folks think about this?